SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverRPC.cpp

```diff
@@ -0,0 +1,451 @@
+#include "EngineClasses/SpatialNetDriverRPC.h"
+#include "EngineClasses/SpatialNetBitReader.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Interop/Connection/SpatialTraceEventBuilder.h"
+#include "Utils/RepLayoutUtils.h"
+
+#include "Interop/RPCs/RPCQueues.h"
+
+using namespace SpatialGDK;
+
+void FRPCMetaData::ComputeSpanId(FName Name, SpatialGDK::SpatialEventTracer& Tracer, SpatialGDK::EntityComponentId EntityComponent,
+								 uint64 RPCId)
+{
+	TArray<FSpatialGDKSpanId> ComponentUpdateSpans = Tracer.GetAndConsumeSpansForComponent(EntityComponent);
+
+	SpanId = Tracer.TraceEvent(
+		FSpatialTraceEventBuilder::CreateReceiveRPC(EventTraceUniqueId::GenerateForNamedRPC(EntityComponent.EntityId, Name, RPCId)),
+		/* Causes */ reinterpret_cast<const Trace_SpanIdType*>(ComponentUpdateSpans.GetData()),
+		/* NumCauses */ ComponentUpdateSpans.Num());
+}
+
+void FRPCPayload::ReadFromSchema(Schema_Object* RPCObject)
+{
+	Offset = Schema_GetUint32(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_OFFSET_ID);
+	Index = Schema_GetUint32(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_INDEX_ID);
+	PayloadData = GetBytesFromSchema(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_PAYLOAD_ID);
+}
+
+void FRPCPayload::WriteToSchema(Schema_Object* RPCObject) const
+{
+	Schema_AddUint32(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_OFFSET_ID, Offset);
+	Schema_AddUint32(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_INDEX_ID, Index);
+	AddBytesToSchema(RPCObject, SpatialConstants::UNREAL_RPC_PAYLOAD_RPC_PAYLOAD_ID, PayloadData.GetData(), PayloadData.Num());
+}
+
+USpatialNetDriverRPC::USpatialNetDriverRPC() {}
+
+USpatialNetDriverRPC::StandardQueue::SentRPCCallback USpatialNetDriverRPC::MakeRPCSentCallback(TArray<UpdateToSend>& OutUpdates)
+{
+	return [&OutUpdates, this](FName Name, Worker_EntityId EntityId, Worker_ComponentId ComponentId, uint64 RPCId,
+							   const FSpatialGDKSpanId& SpanId) {
+		if (EventTracer != nullptr)
+		{
+			EventTraceUniqueId LinearTraceId = EventTraceUniqueId::GenerateForNamedRPC(EntityId, Name, RPCId);
+			FSpatialGDKSpanId NewSpanId = EventTracer->TraceEvent(FSpatialTraceEventBuilder::CreateSendRPC(LinearTraceId),
+																  /* Causes */ SpanId.GetConstId(), /* NumCauses */ 1);
+
+			if (OutUpdates.Num() == 0 || OutUpdates.Last().EntityId != EntityId || OutUpdates.Last().Update.component_id != ComponentId)
```

Comment:
There is a subtle assumption that these are ordered by EntityId then ComponentId, is this a guarantee made somewhere?

Reply:
It's more an assumption that we are treating one entity-component update at a time.
When processing queues, sender and receivers, they will create RPCWritingContext::EntityWrite objects. What I can do is make sure there is only a single one existing associated to a given RPCWritingContext.

Reply:
Ah I see, makes sense then!